### PR TITLE
develop

### PR DESCRIPTION
**Pull request by VKuzmich**

### CODE REVIEW

File: SEED.yaml [hunk 1]
```diff
@@ -0,0 +1,785 @@
+---
```
Comment (yaorlov) [inline SEED.yaml]: Move to `gitignore`

File: application/game_process.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+
+require_relative '../autoload'
+require 'pry'
```
Comment (yaorlov) [inline application/game_process.rb]: remove

File: application/game_process.rb [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    when '/check' then check
+    when '/lose' then lose
+    when '/game' then game
+    when '/win' then win
+    when '/show_hint' then show_hint
+    else Rack::Response.new('Not Found', 404)
+    end
+  end
+
+  # private
```
Comment (yaorlov) [inline application/game_process.rb]: why this line is commented?

File: application/game_process.rb [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  end
+
+  def initialize(env)
+    @request = Rack::Request.new(env)
+  end
+
+  def response
+    case @request.path
+    when '/' then main
+    when '/rules' then Rack::Response.new(render('rules.html.erb'))
```
Comment (yaorlov) [inline application/game_process.rb]: please better move `Rack::Response.new(render('rules.html.erb'))` and `Rack::Response.new(render('statistics.html.erb'))` to methods

File: application/game_process.rb [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  def main
+    return redirect('game') if @request.session.key?(:game)
+
+    Rack::Response.new(render('menu.html.erb'))
+  end
+
+  def play_again
+    @request.session.delete(:game)
+
+    redirect
```
Comment (yaorlov) [inline application/game_process.rb]: better do `redirect('/')`, so it's obvious where you redirect

File: application/game_process.rb [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+      attempts_total: attempts_total,
+      attempts_used: attempts_total - @request.session[:game].attempts,
+      hints_total: hints_total,
+      hints_used: hints_total - @request.session[:game].hints,
+      date: Time.now
+    }
+  end
+
+  def stats
+    return [] unless File.exist?('application/spec/fixtures/seed.yaml')
```
Comment (yaorlov) [inline application/game_process.rb]: why are you using spec `seed.yaml` here?

File: application/game_process.rb [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  def default_setting
+    @request.session.merge!(
+      result: nil,
+      hints: [],
+      save: false
+    )
+  end
+
+  def game_redirect
+    return redirect unless @request.session.key?(:game)
```
Comment (yaorlov) [inline application/game_process.rb]: this logic (`return redirect unless @request.session.key?(:game)`) is repeated a lot, please move it to rack middleware

File: application/game_process.rb [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    return redirect('lose') unless @request.session[:game].attempts.positive?
```
Comment (yaorlov) [inline application/game_process.rb]: `return redirect('lose') if @request.session[:game].attempts.zero?`
prefer `if` over `unless`

File: application/game_process.rb [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    @request.session.delete(:game)
+
+    redirect
+  end
+
+  def start
+    start_redirect || start_initialize
+  end
+
+  def game
```
Comment (yaorlov) [inline application/game_process.rb]: when you remove `return redirect unless @request.session.key?(:game)` from `start_redirect`, then you can move `game_redirect` logic to `game` method and remove `game_redirect` at all
this should be done with all similar methods

File: spec/fixtures/seed.yaml [hunk 10]
```diff
@@ -0,0 +1,6238 @@
+---
```
Comment (yaorlov) [inline spec/fixtures/seed.yaml]: remove this file, it's not used in tests

File: application/game_process.rb (hunk 9 above)
Comment (VKuzmich) [reply application/game_process.rb]: it was done.

File: application/game_process.rb [hunk 11]
```diff
@@ -68,12 +68,16 @@ def start
   end
 
   def game
-    game_redirect || Rack::Response.new(render('game.html.erb'))
+    if @request.session[:game].attempts.zero?
```
Comment (yaorlov) [inline application/game_process.rb]:
```
if @request.session[:game].attempts.zero?
  redirect(URLS[:lose])
else @request.session[:game].win?
  return redirect(URLS[:win])
end
```

File: application/game_process.rb (hunk 9 above)
Comment (yaorlov) [reply application/game_process.rb]: `start_redirect`, `lose_redirect` and similar methods were not removed

File: tmp/rubycritic/application/config/i18n.html [hunk 12]
```diff
@@ -0,0 +1,119 @@
+<!DOCTYPE html>
```
Comment (yaorlov) [inline tmp/rubycritic/application/config/i18n.html]: move all tmp/rubycritic files to gitignore

File: tmp/rubycritic/overview.html [hunk 13]
```diff
@@ -65,8 +65,8 @@ <h3>Overview</h3>
             
 
             <script>
-             var turbulenceData = [{"name":"SpecHelper","x":2,"y":0.0,"rating":"A"},{"name":"GameProcessSpec","x":4,"y":491.97,"rating":"F"},{"name":"DataUtilsSpec","x":3,"y":29.65,"rating":"A"},{"name":"Autoload","x":5,"y":0.0,"rating":"A"},{"name":"TableRow","x":1,"y":9.9,"rating":"A"},{"name":"ReturnRequests","x":2,"y":8.5,"rating":"A"},{"name":"I18n","x":2,"y":0.0,"rating":"A"},{"name":"DataUtils","x":3,"y":9.89,"rating":"A"},{"name":"GameProcess","x":8,"y":208.08,"rating":"D"}];
-             var score = 74.78;
+             var turbulenceData = [{"name":"SpecHelper","x":2,"y":0.0,"rating":"A"},{"name":"GameProcessSpec","x":4,"y":491.97,"rating":"F"},{"name":"DataUtilsSpec","x":3,"y":29.65,"rating":"A"},{"name":"Autoload","x":5,"y":0.0,"rating":"A"},{"name":"TableRow","x":1,"y":9.9,"rating":"A"},{"name":"ReturnRequests","x":2,"y":8.5,"rating":"A"},{"name":"I18n","x":2,"y":0.0,"rating":"A"},{"name":"DataUtils","x":3,"y":9.89,"rating":"A"},{"name":"GameProcess","x":8,"y":205.89,"rating":"D"}];
```
Comment (yaorlov) [inline tmp/rubycritic/overview.html]: gitignore

File: application/middleware/return_requests.rb [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+  def initialize(app, status = 302)
+    @app = app
+    @status = status
+  end
+
+  def call(env)
+    @request = Rack::Request.new(env)
+
+    return [@status, { 'Location' => GameProcess::URLS[:game] }, ['']] if @request.session.key?(:main)
```
Comment (yaorlov) [inline application/middleware/return_requests.rb]: what does `main` mean? maybe you should check if session has a game and a user?
I believe AuthMiddleware is a better name for it

Comment (VKuzmich) [reply application/middleware/return_requests.rb]: Can you explain more what you want me to do?

File: application/game_process.rb (hunk 9 above)
Comment (VKuzmich) [reply application/game_process.rb]: moved start_redirect, lose_redirect and win_redirect methods.

File: application/middleware/auth_middleware.rb [hunk 15]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class AuthMiddleware
```
Comment (yaorlov) [inline application/middleware/auth_middleware.rb]: The goal of this middleware was to remove the duplicated logic from your application: https://github.com/VKuzmich/codebreaker_rack/pull/1/files#r323578648
Several routes check if there is a game in the session. If not, then redirect. You should implement the same logic in the middleware.
Implementation: create an array of paths that should have a game. Then check if the request is going to one of these paths. If yes, do redirect.

File: application/middleware/auth_middleware.rb [hunk 16]
```diff
@@ -9,12 +11,16 @@ def initialize(app, status = 302)
   def call(env)
     @request = Rack::Request.new(env)
 
-    return [@status, { 'Location' => GameProcess::URLS[:game] }, ['']] if authenticated?
+    return [@status, { 'Location' => GameProcess::URLS[:game] }, ['']] if authenticated? && game_links?
```
Comment (yaorlov) [inline application/middleware/auth_middleware.rb]: You already have a redirect to /game in your methods. E.g.: https://github.com/VKuzmich/codebreaker_rack/pull/1/files#diff-b5aef1f120e700160b332ead1ce0f869R80
You should redirect to registration path if it's a game_paths, but the user is not aunthenticated

File: application/middleware/auth_middleware.rb [hunk 17]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class AuthMiddleware
+  GAME_LINKS = ['/check', '/show_hint', '/win', '/lose'].freeze
```
Comment (yaorlov) [inline application/middleware/auth_middleware.rb]: I think GAME_PATHS or AUTH_PATHS is a better naming

File: views/lose.html.erb [hunk 18]
```diff
@@ -0,0 +1,71 @@
+<!DOCTYPE html>
+<html lang="en" class="mdl-js">
+<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+<link href="assets/favicon.ico" rel="icon">
+<link href="assets/css/bootstrap.min.css" rel="stylesheet">
+<link href="assets/css/styles.css" rel="stylesheet">
+<script src="assets/js/jquery.min.js"></script>
```
[the remark this answers is not in the document]
Comment (VKuzmich) [reply views/lose.html.erb]: why do you ask this question?

[the remark this answers is not in the document]
Comment (yaorlov) [reply views/lose.html.erb]: I think you load unnecessary assets. It looks like js assets are not needed here...

[the remark this answers is not in the document]
Comment (VKuzmich) [reply views/lose.html.erb]: I commented this line and ran app. when i losed, the background of page was white. so it is necessary for this page.

File: db/results.yaml [hunk 19]
```diff
@@ -0,0 +1,105 @@
+---
```
Comment (yaorlov) [inline db/results.yaml]: where this file is used? how is it different from seeds file?

**Comment by yaorlov** [comment]
I've got an error after answer submit
Screenshot_20191014_141641

File: application/data_utils.rb [hunk 20]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module DataUtils
+  SEED = 'database/database.yaml'
```
Comment (yaorlov) [inline application/data_utils.rb]: it's not a seed, this constant should be called `DATABASE`

**Pull request merged**